indico/modules/events/persons/client/js/Email.jsx

```diff
@@ -0,0 +1,229 @@
+        params: snakifyKeys(recipientData),
+      });
+    } catch (err) {
+      return handleSubmitError(err);
+    }
+  };
+
+  const renderForm = ({handleSubmit, submitting, submitSucceeded, values, hasValidationErrors}) => (
+    <Form loading={loading} onSubmit={handleSubmit}>
+      {submitSucceeded && (
+        <Message positive>
+          <Translate as={Message.Header}>Your email has been sent.</Translate>
+          <Translate as="p">
+            <Param
+              name="count"
+              value={Object.values(recipientData).reduce((acc, v) => acc + v.length, 0)}
+            />{' '}
+            emails have been sent.
+          </Translate>
+        </Message>
+      )}
+      <Form.Field>
+        <Translate as="label">From</Translate>
+        <FinalDropdown
+          name="fromAddress"
+          scrolling
+          selection
+          options={senders.map(([value, text]) => ({value, text}))}
+        />
+      </Form.Field>
+      <Form.Field>
+        <Translate as="label">Subject</Translate>
+        <FinalField name="subject">
+          {({input: {value, ...rest}}) => (
+            <Input value={preview ? preview.subject : value} disabled={!!preview} {...rest} />
+          )}
+        </FinalField>
+      </Form.Field>
+      <Form.Field>
+        <Translate as="label">Email body</Translate>
+        <FinalField name="body" validate={validators.required}>
+          {({input, meta}) => (
+            <>
+              {meta.touched && (meta.error || meta.submitError) && (
+                <Label basic color="red" pointing="below">
+                  {meta.error || meta.submitError}
+                </Label>
+              )}
+              <TextEditor
+                {...input}
+                value={preview ? preview.body : input.value}
+                onChange={(_, editor) => !disableChange.current && input.onChange(editor.getData())}
+                disabled={!!preview}
+                required
+              />
+            </>
+          )}
+        </FinalField>
+      </Form.Field>
+      <Form.Field>
+        <Translate as="label">Recipients</Translate>
+        <TextArea rows={3} value={recipients.join(', ')} readOnly />
+      </Form.Field>
+      <Form.Field>
+        <Translate as="label">Send copy to me</Translate>
+        <FinalCheckbox name="copyForSender" label={Translate.string('Send copy to me')} toggle />
+        <Translate as="p" className="field-description">
+          Send copy of each email to my mailbox
+        </Translate>
+      </Form.Field>
+      <Translate
+        as={Button}
+        type="submit"
+        disabled={submitting || submitSucceeded || hasValidationErrors}
+        primary
+      >
+        Send
+      </Translate>
+      <Button type="button" active={!!preview} onClick={() => togglePreview(values)}>
+        {!preview ? <Translate>Preview</Translate> : <Translate>Edit</Translate>}
```

Comment:
OK now I see you are using the same field to render the preview. I'm not sure if this is such a good idea - it makes it much harder to properly use out the final-form utils.

What do you think about rendering different stuff in the FinalForm's `render` depending on whether it's preview or edit mode? That way all the form state is kept, and you can use a `FormSpy` for the display mode to get access to the `values` without having to subscribe to value changes in the FinalForm itself (doing so causes large re-renders while typing -> worse performance, right now you don't specify any `subscription` so it's re-rendering on everything)

Reply:
I've tested several approaches before going into the same field, but since I wanted to keep the form data consistent even while in preview, "disabling" the field while still showing its data seemed the more logical approach.

The thing about rendering different stuff is that I was actually repeating the exact same fields (to display the same data) when I actually just wanted to "freeze" them. While re-renders are constantly repeated, I'm assuming the same `FromSpy` could be applied to the button which toggles the preview with the current data.

I think I recall also trying that out, so we can do it.